File: CPP/gameitems.cpp
#include "gameItems.h"
#include "maploader.h"
#include "actorHandler.h"
#include "argHandler.h"
#include "renderer.h"

// =================================================================================================

void CGameItems::Create (void) {

    CreateMap ();
    // the viewer (local player). He controls the projection, i.e. everything is rendered #include "the perspective of the viewer
    m_viewer = actorHandler->CreateViewer ();
    m_viewer->SetupCamera ("viewer", 1.0, CVector (NAN, NAN, NAN), CVector (0, 0, 0));
    m_viewer->ForceRespawn ();
    actorHandler->SetViewer (m_viewer);
    renderer->SetViewer (m_viewer);

    // player reticle
    m_reticle = CReticle ();
    m_reticle.Create ();

    // create some player spheres for testing. These have no function besides being there, being targets, and respawning
    int dummies = argHandler->IntVal ("dummies", 0, 0);
    if (dummies > 0) {
        auto min = [] (auto a, auto b) { (a < b) ? a : b; };
        for (int i = min (dummies, int (actorHandler->m_maxPlayers) - 1); i; i--) {
            CPlayer* dummy = actorHandler->CreatePlayer ();
            dummy->SetType ("dummy");
        }
    }
}


bool CGameItems::CreateMap (void) {
    if (m_map)
        delete m_map;
    m_map = new CMap();
    CString mapName = argHandler->StrVal ("map", 0, "standard.txt");
    if (!CMapLoader (m_map).CreateFromFile (mapName, m_map->m_stringMap)) {
        fprintf (stderr, "Couldn't load map '%s'\n", mapName.Buffer ());
        exit (1);
    }
    return false;
}


bool CGameItems::CreateMap (CList<CString> stringMap, bool isPrepared) {
    if (m_map)
        delete m_map;
    if (!(m_map = new CMap)) {
        fprintf (stderr, "Couldn't load map (out of memory)\n");
        exit (1);
    }
    m_map->m_stringMap = stringMap;
    if (!CMapLoader (m_map).CreateFromMemory (m_map->m_stringMap, isPrepared)) {
        fprintf (stderr, "Couldn't load map\n");
        exit (1);
    }
    return false;
}


CGameItems* gameItems = nullptr;

// =================================================================================================

File: CSharp/gameitems.cs
using System;

// =================================================================================================

public class GameItems
{
    public Map m_map;
    public Viewer m_viewer;
    public Reticle m_reticle;

    public GameItems() { }

    public void Destroy()
    {
        m_map.Destroy();
        m_reticle.Destroy();
        m_map = null;
        m_reticle = null;
    }

    public void Cleanup()
    {
        Globals.actorHandler.Cleanup();
    }

    public float ViewerDistance(Vector p)
    {
        return m_map.Distance(m_viewer.GetPosition(), p);
    }

    public void Create()
    {

        CreateMap();
        // the viewer (local player). He controls the projection, i.e. everything is rendered #include "the perspective of the viewer
        m_viewer = Globals.actorHandler.CreateViewer();
        m_viewer.SetupCamera("viewer", 1.0f, new Vector(float.NaN, float.NaN, float.NaN), new Vector(0, 0, 0));
        m_viewer.ForceRespawn();
        Globals.actorHandler.SetViewer(m_viewer);
        Globals.renderer.SetViewer(m_viewer);

        // player reticle
        m_reticle = new Reticle();
        m_reticle.Create();

        // create some player spheres for testing. These have no function besides being there, being targets, and respawning
        int dummies = Globals.argHandler.IntVal("dummies", 0, 0);
        if (dummies > 0)
        {
            for (int i = Math.Min(dummies, (int)Globals.actorHandler.m_maxPlayers - 1); i > 0; i--)
            {
                Player dummy = Globals.actorHandler.CreatePlayer();
                dummy.SetType("dummy");
            }
        }
    }


    public bool CreateMap()
    {
        m_map = new Map();
        string mapName = Globals.argHandler.StrVal("map", 0, "standard.txt");
        MapLoader mapLoader = new MapLoader(m_map);
        m_map.m_stringMap = mapLoader.CreateFromFile(mapName);
        if (m_map.m_stringMap == null)
        {
            Console.Error.WriteLine("Couldn't load map '{0}'", mapName);
            System.Environment.Exit(1);
        }
        return true;
    }


    public bool CreateMap(string[] stringMap, bool isPrepared = false)
    {
        if (m_map != null)
            m_map.Destroy();
        m_map = new Map();
        if (m_map == null)
        {
            Console.Error.WriteLine("Couldn't load map (out of memory)");
            System.Environment.Exit(1);
        }
        m_map.m_stringMap = stringMap;
        MapLoader mapLoader = new MapLoader(m_map);
        if (!mapLoader.CreateFromMemory(m_map.m_stringMap, isPrepared))
        {
            Console.Error.WriteLine("Couldn't load map");
            System.Environment.Exit(1);
        }
        return true;
    }

}

// =================================================================================================

